code/obj/submachine/seed.dm

```diff
@@ -1084,26 +1084,111 @@
-		if (!src.can_vend)
-			dat+= "<u>Unit currently out of charge. Please wait.</u><br>"
-		dat += "<br>"
+	proc/getPlantImg(var/datum/plant/plant)
+		if(!istype(plant))
+			return
+		var/path = plant.type
+		var/img = plant_base64_cache[path]
+		if(!img)
+			var/icon/result_icon
+			if(plant.crop)
+				var/atom/crop = plant.crop
+				result_icon = icon(initial(crop.icon), initial(crop.icon_state))
+			else if(plant.plant_icon)
+				if(plant.sprite)
+					result_icon = icon(plant.plant_icon, "[plant.sprite]-G4", frame=1)
+				else if(plant.override_icon_state)
+					result_icon = icon(plant.plant_icon, "[plant.override_icon_state]-G4", frame=1)
+				else
+					result_icon = icon(plant.plant_icon, "[plant.name]-G4", frame=1)
+
+			if(result_icon)
+				img = icon2base64(result_icon)
+				plant_base64_cache[path] = img
+		return img
+
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if(!ui)
+			ui = new(user, src, "SeedFabricator", src.name)
+			ui.open()
+
+	ui_data(mob/user)
+		. = list()
+		.["seedCount"] = src.seedcount
+		.["canVend"] = src.can_vend
+		.["isWorking"] = src.working
+
+	ui_static_data(mob/user)
+		. = list()
+
+		.["maxSeed"] = src.maxseed
+		.["name"] = src.name
+
+		// Start with associative list, where each key is a seed category
+		var/list/categories = list()
 		for(var/datum/plant/A in hydro_controls.vendable_plants)
 			if (A.vending == 1 || src.hacked)
-				if (!src.category || (src.category == A.category))
-					dat += "<b>[A.name]</b>: <A href='?src=\ref[src];disp=\ref[A]'>(VEND)</A><br>"
+				if (!categories[A.category])
+					categories[A.category] = list()
+				categories[A.category] += list(list(
+					name = A.name,
+					path = A.type,
+					img = getPlantImg(A)
+				))
+		// Convert to non-associative list holding each category
+		var/list/categoriesArray = list()
+		for(var/category_name in categories)
+			var/category = categories[category_name]
+			categoriesArray += list(list(
+				name = category_name,
+				seeds = category
+			))
+		.["seedCategories"] = categoriesArray
+
+
+	ui_act(action, params)
+		. = ..()
+		if(. || action != "disp" || !src.can_vend || !src.working)
+			return
+		var/datum/plant/I = locate(text2path(params["path"])) in src.available
+
+		if (!istype(I))
+			return
+
+		if(!I.vending)
+			trigger_anti_cheat(usr, "tried to href exploit vend forbidden seed [I] on [src]")
+			return
+
+		var/vend = clamp(params["amount"], 1, 10)
+
+		while(vend > 0)
+			var/obj/item/seed/S
+			if (I.unique_seed)
+				S = new I.unique_seed
+				S.set_loc(src.loc)
+			else
+				S = new /obj/item/seed
+				S.set_loc(src.loc)
+				S.removecolor()
+			S.generic_seed_setup(I)
+			vend--
+			src.seedcount++
+		SPAWN_DBG(0)
+			for(var/obj/item/seed/S in src.contents) S.set_loc(src.loc)
+		if(src.seedcount >= src.maxseed)
+			src.can_vend = 0
+			SPAWN_DBG(10 SECONDS)
+				src.can_vend = 1
+				src.seedcount = 0
```

Comment:
da fuk is this code

Reply:
Yeah, the seed fabricator has a weird way of recharging, how do you recommend I should change it?

Reply:
no like, the spawn and for loop